src/python/pants/engine/internals/graph.py

```diff
@@ -135,37 +134,74 @@ async def resolve_targets_with_origins(
 # -----------------------------------------------------------------------------------------------
 
 
-@rule
-async def transitive_target(wrapped_root: WrappedTarget) -> TransitiveTarget:
-    root = wrapped_root.target
-    if not root.has_field(Dependencies):
-        return TransitiveTarget(root, ())
-    dependency_addresses = await Get(Addresses, DependenciesRequest(root[Dependencies]))
-    dependencies = await MultiGet(Get(TransitiveTarget, Address, d) for d in dependency_addresses)
-    return TransitiveTarget(root, dependencies)
+class CycleException(Exception):
+    def __init__(self, subject: Address, path: Tuple[Address, ...]) -> None:
+        path_string = "\n".join((f"-> {a}" if a == subject else f"   {a}") for a in path)
+        super().__init__(f"Dependency graph contained a cycle:\n{path_string}")
```

Comment:
Great! This is close to fixing https://github.com/pantsbuild/pants/issues/10390.

We possibly want to add a suggestion on how to fix this? It might not be obvious that you should consider creating more precise targets. Totally fine to punt on this, though, because it depends in part on a) how we allow you to opt into tolerance, and b) what happens to generated subtargets.

Reply:
Yep, should probably punt.